backend/db.js

```diff
@@ -0,0 +1,106 @@
+import { DataSource } from 'apollo-datasource'
+import crypto from 'crypto'
+
+export class Post {
+    constructor(data) {
+        this.id = crypto.randomBytes(16).toString('hex')
+        Object.assign(this, data)
+    }
+}
+
+export class User {
+    constructor(data) {
+        this.id = crypto.randomBytes(16).toString('hex')
+        Object.assign(this, data)
+    }
+}
+
+export class InMemoryDataSource extends DataSource {
+    constructor() {
+        super()
+        this.posts = []
+        this.users = []
+        this.initDummyUsers()
+        this.initDummyPosts()
+    }
+
+    initialize(...args) {
+        //Warum wird diese Methode so oft ausgeführt?
+    }
+
+    createPost(data) {
+        const author = this.users.find(user => user.name === data.post.author.name)
+        const newPost = new Post({
+            title: data.post.title,
+            votes: [],
+            author: author
+        })
+        this.posts.push(newPost)
+        const newUsers = this.users.map(user => user.id === newPost.author.id ? this.addPostToAuthor(newPost, user) : user)
+        this.users = [...newUsers]
+
+        return newPost
+    }
+
+    upvotePost(data) {
+        const voter = this.users.find(user => user.name === data.voter.name)
+        const newPosts = this.posts.map(post => post.id === data.id && post.votes.find(user => user.name === voter.name) === undefined ? this.addVoter(post, voter) : post)
+        this.posts = [...newPosts]
+
+        return this.posts.find(post => post.id === data.id)
+    }
+
+    getVotes(postId) {
+        return this.posts.find(post => post.id === postId).votes.length
+    }
```

Comment:
```suggestion
    getPost(postId) {
        return this.posts.find(post => post.id === postId)
    }
```

And in your resolvers:
```js
export default {
  Query: //,
  Mutation: //,
  Post: {
    votes: (post) => post.votes.length
  }
}
```

Separation of concerns + you cannot forget to provide `votes` for a post anymore.